src/tools/pkcs11-tool.c

```diff
@@ -1223,9 +1223,12 @@ int main(int argc, char * argv[])
 
 	if (do_sign || do_derive || do_decrypt) {
 		if (!find_object(session, CKO_PRIVATE_KEY, &object,
+					opt_object_id_len ? opt_object_id : NULL,
+					opt_object_id_len, 0) &&
+			!find_object(session, CKO_SECRET_KEY, &object,
 					opt_object_id_len ? opt_object_id : NULL,
 					opt_object_id_len, 0))
-			util_fatal("Private key not found");
+			util_fatal("Private key nor secret key found");
```

Comment:
You should probably split the above conditions to match only your intended operation (Sign). Although decrypt might work also for some secret keys and mechanisms, I am not sure about the derive operation.

Reply:
Not to mention that in cryptography there's no signing with symmetric keys, only MAC'ing.

It's good to be able to compute keyed hash (e.g., HMAC) of an object, but I wonder what's the actual use case is.

Reply:
> Not to mention that in cryptography there's no signing with symmetric keys, only MAC'ing.
> 

Which is effectively and practically a symmetric signature and is referred to that in both the TCG TPM Architectural docs as well as practically through PKCS11 via the C_Sign and C_Verify operations as well as OpenSSL's C API of EVP_Signa nd EVP_Verify interfaces.

> It's good to be able to compute keyed hash (e.g., HMAC) of an object, but I wonder what's the actual use case is.

Theirs a lot of specific use cases for an HMAC, but the general use case is always the same, detect authenticity and tampering.

Reply:
> You should probably split the above conditions to match only your intended operation (Sign). Although decrypt might work also for some secret keys and mechanisms, I am not sure about the derive operation.

I split it up, trying to keep the patch as small as possible to accomplish the goal, HMAC sign and verify.